library/core/src/alloc/layout.rs

```diff
@@ -172,27 +172,33 @@ impl Layout {
     /// allocate backing structure for `T` (which could be a trait
     /// or other unsized type like a slice).
     ///
+    /// This is an unchecked version of [`for_value`][Self::for_value]
+    /// which takes a raw pointer instead of a reference.
+    ///
     /// # Safety
     ///
-    /// This function is only safe to call if the following conditions hold:
-    ///
-    /// - If `T` is `Sized`, this function is always safe to call.
-    /// - If the unsized tail of `T` is:
-    ///     - a [slice], then the length of the slice tail must be an initialized
-    ///       integer, and the size of the *entire value*
-    ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
-    ///     - a [trait object], then the vtable part of the pointer must point
-    ///       to a valid vtable for the type `T` acquired by an unsizing coercion,
-    ///       and the size of the *entire value*
-    ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
-    ///     - an (unstable) [extern type], then this function is always safe to
-    ///       call, but may panic or otherwise return the wrong value, as the
-    ///       extern type's layout is not known. This is the same behavior as
-    ///       [`Layout::for_value`] on a reference to an extern type tail.
-    ///     - otherwise, it is conservatively not allowed to call this function.
+    /// The provided (possibly wide) pointer must describe a valid value layout.
+    /// Specifically:
+    ///
+    /// - If `T` is a `Sized` type, this function is always safe to call and is
+    ///   equivalent to [`size_of::<T>()`][size_of]. The pointer is unused.
+    /// - If the unsized tail of `T` is a [slice], then the size of the *entire
+    ///   value* (statically sized prefix plus dynamic tail) must fit in `isize`.
+    ///   The pointer may be dangling or null; only the slice length is used.
+    /// - If the unsized tail of `T` is a [trait object], then the wide pointer
+    ///   metadata (the vtable reference) must originate from an unsizing coercion,
+    ///   and the size of the *entire value* (statically sized prefix plus dynamic
+    ///   tail) must fit in `isize`. The pointer may be dangling or null; only the
+    ///   wide pointer metadata is used.
+    /// - For any other unsized tail kind (for example, unstable [extern types]),
+    ///   it is currently *not permitted* to call this function. Unknown unsized
+    ///   tail kinds may impose arbitrary requirements unknowable to current code.
```

Comment:
Not permitted means... UB?

Reply:
The "not allowed" wording comes from the existing docs (which I also wrote). The gist is "the safety condition depends on the pointee type and if it's not one of these documented kinds you cannot satisfy safety conditions which you don't know" but that's a bit of a mouthful. I also would prefer not to say "if `size_of_val(&*ptr)` is safe, `size_of_val_raw(&*ptr)` is safe," because that's not *necessarily* true, especially if we take the option of making `?Sized` mean `?DynSized` and have `size_of_val` panic for types where size can't be determined, instead of the more complicated path of adding support for types more unsized than `?Sized`[^1].

[^1]: Which actually works mostly fine for any type which has a default `Sized` bound and supports the `?Sized` bound, but quickly gets (maybe impossibly) messy for ones that don't have a default `Sized` bound applied, such as `Self` in trait default method bodies and generic types' associated type projections (when the associated type is relaxed further than `?Sized`, which must happen[^2] because `Deref::Target` needs to be fully unbound for anything to work).

[^2]: *Perhaps* if we develop a solution for migrating to GAT-ified traits, that same solution could be used for only sometimes implying a `MetaSized` bound on the associated type. My implementation attempt didn't inspire hope.

So I suppose yes, it would be (library) undefined behavior, but *right now*, actually it would *non*-behavior, as there are no such types.

Reply:
For all I can tell "not permitted" means "precondition: False" means `unreachable_unchecked()`. So it seems more honest to just say that it's UB, possibly followed by a remark that on *current* stable Rust, no such type exists.

FWIW in Miri this will panic.

Reply:
I switched it to say it's UB, but didn't add any note that current stable doesn't have such tails. That the example is unstable and it does talk about "unknown unsized tail kinds" does imply that it's about future compatibility, though.